share/man/man1/brew.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "BREW" "1" "November 2010" "Homebrew" "brew"
.
.SH "NAME"
\fBbrew\fR \- The missing package manager for OS X
.
.SH "SYNOPSIS"
\fBbrew\fR [\-\-verbose|\-v] command [options] [formula] \.\.\.
.
.br
\fBbrew\fR [\-\-version|\-v]
.
.SH "DESCRIPTION"
Homebrew is the easiest and most flexible way to install the UNIX tools Apple didn\'t include with OS X\.
.
.SH "OPTIONS"
.
.TP
\fB\-v\fR, \fB\-\-verbose\fR command [options] [formula] \.\.\.
Prints extra, comand\-specific debugging information\. Note that \fBbrew \-v\fR by itself is the same as \fBbrew \-\-version\fR\.
.
.SH "COMMANDS"
.
.TP
\fB\-v\fR, \fB\-\-version\fR
Prints the version number of brew to standard error and exits\.
.
.TP
\fB\-\-prefix\fR [\fIformula\fR]
Displays the install path for Homebrew\. If \fIformula\fR is given, display location in the cellar where that package is or would be installed\.
.
.TP
\fB\-\-cache\fR [\fIformula\fR]
Displays the path Homebrew uses to cache downloads\. If \fIformula\fR is given, display the file or folder used to cache that specific package\.
.
.TP
\fB\-\-config\fR
Shows Homebrew and system configuration useful for debugging\. If you file a bug report, you will likely be asked for this information if you do not provide it\.
.
.TP
\fB\-S\fR, \fBsearch\fR \fItext\fR|/\fItext\fR/
Performs a substring search of formula names for \fItext\fR\. If \fItext\fR is surrounded with slashes, then it is interpreted as a regular expression\. If no search term is given, all available formula are displayed\.
.
.TP
\fBinstall [\-\-force] [\-\-debug] [\-\-use\-llvm] [\-\-ignore\-dependencies] [\-\-HEAD]\fR \fIformula\fR
Installs \fIformula\fR\.
.
.IP
\fIformula\fR is usually the name of the formula to install, but may also be the URL for an arbitrary formula\.
.
.IP
If \fB\-\-force\fR is passed, will install \fIformula\fR even if it is already installed\. This can be used to re\-install a formula without removing it first\.
.
.IP
If \fB\-\-debug\fR is passed and brewing fails, opens a shell inside the temporary folder used for compiling\.
.
.IP
If \fB\-\-use\-llvm\fR is passed, attempt to compile using the LLVM front\-end to GCC\. \fINOTE\fR: Not all formulae will build with LLVM\.
.
.IP
If \fB\-\-ignore\-dependencies\fR is passed, skip installing any dependencies of any kind\. If they are not already present, the formula will probably fail to install\.
.
.IP
If \fB\-\-HEAD\fR is passed, and \fIformula\fR defines it, install the HEAD version, aka master, trunk, unstable, dev\.
.
.IP
To install a newer version of HEAD use \fBbrew rm <foo> && brew install \-\-HEAD <foo>\fR or \fBbrew \-\-force \-\-HEAD <foo>\fR\.
.
.TP
\fBinstall \-\-interactive [\-\-git]\fR \fIformula\fR
Downloads and patches \fIformula\fR, then opens a shell\. This allows the user to run \fB\./configure \-\-help\fR and otherwise determine how to turn the software package into a Homebrew formula\.
.
.IP
If \fB\-\-git\fR is passed, Homebrew will create a Git repository, useful for creating patches to the software\.
.
.TP
\fBlist\fR
Lists all installed formulae\.
.
.TP
\fBlist\fR \fIformula\fR
Lists the installed files for \fIformula\fR\.
.
.TP
\fBinfo\fR \fIformula\fR
Gives all available information for \fIformula\fR\.
.
.TP
\fBinfo \-\-github\fR \fIformula\fR
Opens a browser to the GitHub History page for formula \fIformula\fR\.
.
.TP
\fBinfo \-\-all\fR
Summarises all installed packages; this is intended to be used by higher\-level tools\.
.
.TP
\fBinfo\fR \fIURL\fR
Prints the name and version that will be detected for \fIURL\fR; only http:// URLs supported for now\.
.
.TP
\fBhome\fR
Opens a browser to Homebrew\'s own homepage\.
.
.TP
\fBhome\fR \fIformula\fR
Opens a browser to \fIformula\fR\'s homepage\.
.
.TP
\fBrm\fR, \fBremove [\-\-force]\fR \fIformula\fR
Uninstalls \fIformula\fR\.
.
.IP
If \fB\-\-force\fR is passed, and there are multiple versions of \fIformula\fR installed, deletes all installed versions\.
.
.TP
\fBcreate [\-\-cache]\fR \fIURL\fR
Generates a formula for the downloadable file at \fIURL\fR and opens it in $EDITOR\. Homebrew will attempt to automatically derive the formula name and version, if it fails, you\'ll have to make your own template\. I suggest copying wget\'s\.
.
.IP
If \fB\-\-cache\fR is passed, Homebrew will download the \fIURL\fR to the cache and add the MD5 to the formula for you\.
.
.TP
\fBcreate \-\-macports\fR|\fB\-\-fink\fR \fIformula\fR
Opens a browser to the MacPorts or Fink package search page, so you can see how they do \fIformula\fR\.
.
.TP
\fBedit\fR \fIformula\fR
Opens the formula in $EDITOR\.
.
.TP
\fBedit\fR
Opens all of Homebrew for editing in TextMate\.
.
.TP
\fBln\fR, \fBlink\fR \fIformula\fR
Symlinks all of \fIformula\fR\'s installed files into the Homebrew prefix\. This is done automatically when you install formula\. It is useful for DIY installation, or in cases where you want to swap out different versions of the same package that you have installed at the same time\.
.
.TP
\fBunlink\fR \fIformula\fR
Unsymlinks \fIformula\fR from the Homebrew prefix\.
.
.TP
\fBprune\fR
Removes dead symlinks from the Homebrew prefix\. This is generally not needed\. However, it can be useful if you are doing DIY installations\.
.
.TP
\fBoutdated\fR
Shows formula that have an updated version available\.
.
.TP
\fBdeps [\-\-1]\fR \fIformula\fR
Shows \fIformula\fR\'s dependencies\.
.
.IP
If \fB\-\-1\fR is passed, only show dependencies one level down, instead of recursing\.
.
.TP
\fBuses [\-\-installed]\fR \fIformula\fR
Shows the formulas that specify \fIformula\fR as a dependency\. The list is not recursive; only one level of dependencies is resolved\.
.
.IP
If \fB\-\-installed\fR is passed, only lists installed formulae\.
.
.TP
\fBdoctor\fR
Checks your system for potential problems\.
.
.TP
\fBcat\fR \fIformula\fR
Displays the source to \fIformula\fR\.
.
.TP
\fBcleanup\fR [\fIformula\fR]
For all installed or specific formulae, remove any older versions from the cellar\.
.
.TP
\fBupdate\fR
Using Git, fetches the newest version of Homebrew from the GitHub repository\.
.
.TP
\fBlog [git\-options]\fR \fIformula\fR \.\.\.
Shows the git log for the given formulae\. Options that \fBgit log\fR recognizes can be passed before the formula list\.
.
.TP
\fBfetch [\-\-force] [\-v] [\-\-HEAD]\fR \fIformula\fR
Downloads the tarball or checks out from VCS for the given \fIformula\fR\. For tarballs, also prints MD5 and SHA1 checksums\.
.
.IP
If \fB\-\-HEAD\fR is passed, download the HEAD version of \fIformula\fR instead\. \fB\-v\fR may also be passed to make the VCS checkout verbose, useful for seeing if an existing HEAD cache has been updated\.
.
.IP
If \fB\-\-force\fR is passed, remove a previously cached version and re\-fetch\.
.
.SH "EXTERNAL COMMANDS"
Homebrew allows external commands to be defined by putting a +x file named \fBbrew\-<cmdname>\fR or \fBbrew\-<cmdname>\.rb\fR on the PATH\. This will cause Homebrew to recognize \fBbrew cmdname\fR\.
.
.P
Some external commands are shipped with Homebrew, and enabled by default\.
.
.TP
\fBaudit [\-\-warn]\fR
Checks all formulae for Homebrew coding style violations\. This should be run before submitting a new formula for inclusion\.
.
.IP
If \fB\-\-warn\fR is passed, performs additional stricter checks that may not need to be fixed before submitting\.
.
.TP
\fBoptions\fR \fIformula\fR
Displays install options specific to \fIformula\fR\.
.
.TP
\fBman\fR
Regenerates this man page from source\.
.
.IP
\fINOTE\fR: Requires \fBronn\fR \fIhttp://rtomayko\.github\.com/ronn/\fR\.
.
.TP
\fBmissing\fR
Checks all installed formuale for missing dependencies\.
.
.TP
\fBserver\fR
Starts a local webserver with an app that lets you browse available formulae, similar to \fBgem server\fR\.
.
.IP
\fINOTE\fR: Requires \fBsinatra\fR \fIhttp://www\.sinatrarb\.com/\fR\.
.
.TP
\fBtest\fR \fIformula\fR
A few formulae provide a test method\. \fBbrew test <formula>\fR runs this test method\. There is no standard output or return code, but it should generally indicate to the user if something is wrong with the installed formula\.
.
.IP
Example: \fBbrew install jruby && brew test jruby\fR
.
.SH "ENVIRONMENT"
.
.TP
HOMEBREW_DEBUG
If set, instructs Homebrew to always assume \fB\-\-debug\fR when running commands\.
.
.TP
HOMEBREW_DEBUG_INSTALL
When \fBbrew install \-d\fR or \fBbrew install \-i\fR drops into a shell, \fBHOMEBREW_DEBUG_INSTALL\fR will be set to the name of the formula being brewed\.
.
.TP
HOMEBREW_DEBUG_PREFIX
When \fBbrew install \-d\fR or \fBbrew install \-i\fR drops into a shell, \fBHOMEBREW_DEBUG__PREFIX\fR will be set to the target prefix in the Cellar of the formula being brewed\.
.
.TP
HOMEBREW_EDITOR
If set, Homebrew will use this editor when editing a single formula, or several formulae in the same folder\.
.
.IP
\fINOTE\fR: \fBbrew edit\fR will open all of Homebrew as discontinuous files and folders\. TextMate can handle this correctly in project mode, but many editors will do strange things in this case\.
.
.TP
HOMEBREW_KEEP_INFO
If set, Homebrew will not remove files from share/info, allowing them to be linked from the Cellar\.
.
.TP
HOMEBREW_MASTER_REPO, HOMEBREW_MASTER_BRANCH
These default to "mxcl" and "master" respectively, and determine which repository to pull from when doing \fBbrew update\fR\.
.
.IP
These can be set to an alternate repository on GitHub to force updates to come from an alternate repository\.
.
.TP
HOMEBREW_SVN
When exporting from Subversion, Homebrew will use \fBHOMEBREW_SVN\fR if set, a Homebrew\-built Subversion if installed, or the system\-provided binary\.
.
.IP
Set this to force Homebrew to use a particular svn binary\.
.
.TP
HOMEBREW_TEMP
If set, instructs Homebrew to use \fBHOMEBREW_TEMP\fR as the temporary folder for building packages\. This may be needed if your system temp folder and Homebrew Prefix are on different volumes, as OS X has trouble moving symlinks across volumes when the target does not yet exist\.
.
.IP
This issue typically occurs when using FileVault (or certain custom SSD configurations\.)
.
.TP
HOMEBREW_USE_LLVM
If set, instructs Homebrew to use the LLVM front\-ends to the GCC compilers\.
.
.IP
\fINOTE\fR: Not all formulae will build correctly under LLVM\.
.
.TP
HOMEBREW_VERBOSE
If set, instructs Homebrew to always assume \fB\-\-verbose\fR when running commands\.
.
.SH "USING HOMEBREW BEHIND A PROXY"
Homebrew uses several commands for downloading files (e\.g\. curl, git, svn)\. Many of these tools can download via a proxy\. It\'s common for these tools to read proxy parameters from environment variables\.
.
.P
For the majority of cases setting \fBhttp_proxy\fR is enough\. You can set this in your shell profile, or you can use it before a brew command:
.
.IP "" 4
.
.nf

http_proxy=http://<host>:<port> brew install foo
.
.fi
.
.IP "" 0
.
.P
If your proxy requires authentication:
.
.IP "" 4
.
.nf

http_proxy=http://<user>:<password>@<host>:<port> brew install foo
.
.fi
.
.IP "" 0
.
.SH "SEE ALSO"
Homebrew Wiki: http://wiki\.github\.com/mxcl/homebrew/
.
.SH "AUTHORS"
Max Howell, a splendid chap\.
.
.SH "BUGS"
See Issues on GitHub: http://github\.com/mxcl/homebrew/issues


Library/Homebrew/brew.h.rb
FORMULA_META_FILES = %w[README README.md ChangeLog COPYING LICENSE LICENCE COPYRIGHT AUTHORS]
PLEASE_REPORT_BUG = "#{Tty.white}Please report this bug at #{Tty.em}http://github.com/mxcl/homebrew/issues#{Tty.reset}"

def issues_for_formula name
  # bit basic as depends on the issue at github having the exact name of the
  # formula in it. Which for stuff like objective-caml is unlikely. So we
  # really should search for aliases too.

  name = f.name if Formula === name

  require 'open-uri'
  require 'yaml'

  issues = []

  open("http://github.com/api/v2/yaml/issues/search/mxcl/homebrew/open/"+name) do |f|
    YAML::load(f.read)['issues'].each do |issue|
      issues << 'http://github.com/mxcl/homebrew/issues/#issue/%s' % issue['number']
    end
  end

  issues
rescue
  []
end

def macports_or_fink_installed?
  # See these issues for some history:
  # http://github.com/mxcl/homebrew/issues/#issue/13
  # http://github.com/mxcl/homebrew/issues/#issue/41
  # http://github.com/mxcl/homebrew/issues/#issue/48

  %w[port fink].each do |ponk|
    path = `/usr/bin/which -s #{ponk}`
    return ponk unless path.empty?
  end

  # we do the above check because macports can be relocated and fink may be
  # able to be relocated in the future. This following check is because if
  # fink and macports are not in the PATH but are still installed it can
  # *still* break the build -- because some build scripts hardcode these paths:
  %w[/sw/bin/fink /opt/local/bin/port].each do |ponk|
    return ponk if File.exist? ponk
  end

  # finally, sometimes people make their MacPorts or Fink read-only so they
  # can quickly test Homebrew out, but still in theory obey the README's 
  # advise to rename the root directory. This doesn't work, many build scripts
  # error out when they try to read from these now unreadable directories.
  %w[/sw /opt/local].each do |path|
    path = Pathname.new(path)
    return path if path.exist? and not path.readable?
  end
  
  false
end

# For backwards compatibility; will be removed in Homebrew 0.9
def outdated_brews
  return HOMEBREW_CELLAR.outdated_brews
end

# For backwards compatibility; will be removed in Homebrew 0.9
def prune
  return HOMEBREW_PREFIX.prune
end

def search_brews text
  require "formula"

  return Formula.names if text.to_s.empty?

  rx = if text =~ %r{^/(.*)/$}
    Regexp.new($1)
  else
    /.*#{Regexp.escape text}.*/i
  end

  aliases = Formula.aliases
  results = (Formula.names+aliases).grep rx

  # Filter out aliases when the full name was also found
  results.reject do |alias_name|
    if aliases.include? alias_name
      results.include? Formula.resolve_alias(alias_name)
    end
  end
end

def _compiler_recommendation build, recommended
  message = (!build.nil? && build < recommended) ? "(#{recommended} or newer recommended)" : ""
  return build, message
end

def dump_config
  require 'hardware'
  sha = `cd #{HOMEBREW_REPOSITORY} && git rev-parse --verify HEAD 2> /dev/null`.chomp
  sha = "(none)" if sha.empty?
  bits = Hardware.bits
  cores = Hardware.cores_as_words
  kernel_arch = `uname -m`.chomp
  system_ruby = Pathname.new("/usr/bin/ruby")

  llvm,   llvm_msg   = _compiler_recommendation llvm_build,   RECOMMENDED_LLVM
  gcc_42, gcc_42_msg = _compiler_recommendation gcc_42_build, RECOMMENDED_GCC_42
  gcc_40, gcc_40_msg = _compiler_recommendation gcc_40_build, RECOMMENDED_GCC_40
  xcode = xcode_version || "?"

  puts <<-EOS
HOMEBREW_VERSION: #{HOMEBREW_VERSION}
HEAD: #{sha}
HOMEBREW_PREFIX: #{HOMEBREW_PREFIX}
HOMEBREW_CELLAR: #{HOMEBREW_CELLAR}
HOMEBREW_REPOSITORY: #{HOMEBREW_REPOSITORY}
HOMEBREW_LIBRARY_PATH: #{HOMEBREW_LIBRARY_PATH}
Hardware: #{cores}-core #{bits}-bit #{Hardware.intel_family}
OS X: #{MACOS_FULL_VERSION}
Kernel Architecture: #{kernel_arch}
Ruby: #{RUBY_VERSION}-#{RUBY_PATCHLEVEL}
/usr/bin/ruby => #{system_ruby.realpath}
Xcode: #{xcode}
GCC-4.0: #{gcc_40 ? "build #{gcc_40}" : "N/A"} #{gcc_42_msg}
GCC-4.2: #{gcc_42 ? "build #{gcc_42}" : "N/A"} #{gcc_40_msg}
LLVM: #{llvm ? "build #{llvm}" : "N/A" } #{llvm_msg}
MacPorts or Fink? #{macports_or_fink_installed?}
X11 installed? #{x11_installed?}
EOS
end
